### ISSUE TITLE
Android `RecreationAttempt` on Activity reopen 

### ISSUE THREAD

**Issue by SnowyCoder** [rust-windowing/winit#3325]
When an android app calls `event_loop.exit()` the application exits, but if the application is opened again without closing it in the "recent apps" the EventLoop creation will fail with a `RecreationAttempt` error.

## Steps to reproduce
- clone https://github.com/SnowyCoder/winit-bug-reopen-example
- run in an android phone
- press back button
- reopen app

## Reported error
```called `Result::unwrap()` on an `Err` value: RecreationAttempt```

## Repo explaination
The repo linked in the "Steps to reproduce" section is a modified version of the one linked in the README android section.
I just updated the dependencies and linked `BrowserBack` to `event_loop.exit()` (that sems to be the logical key invoked when the android back button is pressed).

## Comment
The general problem seems to be that `android_main` can be called more than once without unloading and reloading the native library (ex. if we use an application with two interacting activities, say LoginActivity and GameActivity).

**Comment by daxpedda** [comment]
See https://github.com/rust-windowing/winit/pull/2344#issuecomment-1166662533, which points out how this can happen on Android.

But generally speaking this is the intended behavior by Winit, you just have to figure out how to solve this instead of trying to recreate the event loop.

Let us know if this solves your problem!

**Comment by SnowyCoder** [comment]
I read that conversation and it doesn't really say how to solve the issue, I've tried various approaches but haven't yet had success. Is there any example Android app that has this bug solved?

Here are my attempts:
1. First thread creates the event loop, saves it in a const and next threads use that loop.
    Cannot work since event loops cannot be shared across threads
2. Since that conversation talks about not creating the loop in the fake main function, create a common thread that manages the event loop and when a new invocation occurs it calls into the common thread
  Does not work since the new thread does not have an attached looper
3. Let the first thread create the event loop and keep it alive  as a  "common thread" like in the second approach, reusing that event loop for the next invocations.
  Since the main thread never exits the app stops responding (I'm blocking the UI thread?)

I think that the "best" approach would be to augment the second point, spawn a new common thread that manages the event loop and manually create a looper out of it. But it involves nasty low-level JNI calls and I'm not even sure if it would really work.
If we let android recreate its event loop once it exits is, to me, the most elegant option.
(tested it commenting the `RecreationAttempt` line, doesn't show any sign of error).

TL;DR: I tried, but I have no idea on how to get it working without recreating the event loop or calling JNI stuff.

**Comment by MarijnS95** [comment]
Per the linked comment, after previously muddling along with these issues in `ndk-glue` for a while, none of those design issues were addressed when `android-activity` was created :weary: 

Blindly recreating the loop when a _new activity is created_ seems appalling. Consider activities to be analogous to `Window`s. Your scenario concerns "closing" one activity (I think) and then later having a new activity opened. However, it is also possible to have _multiple activities opened at the same time_.

Fwiw I think exiting the loop and returning out of `android_main()` causes your activity to be closed because of https://github.com/rust-mobile/android-activity/issues/67 / https://github.com/rust-mobile/android-activity/pull/81. Also notice that the callback function from Android here is called `ANativeActivity_onCreate()` which **is not** analogous to something called `main()`, yet pretends to be used as such.

---

We'll need to be more thoughtful how to handle that across the `winit` and `android-activity` bindings. You'll see the `ANativeActivity_onCreate()` callback in `android-activity` is already creating a thread to run your `android_main()` and the event loop in. That's likely the thread that will have to stay alive to process incoming window creation requests through `ANativeActivity_onCreate()`, until the process exits.

Then, we'll likely need a "New Window Created" event in `winit` to process this (and make the `Window` impl for Android a less-empty shell), and also improve the `Suspended`/`Resumed` events to operate per-surface (should be a 1:1 mapping with an activity) instead of just globally.

This means that creating a new window in `winit` will no longer do what you expect, and we'll have to craft some careful bindings around it.

**Comment by MarijnS95** [comment]
https://developer.android.com/guide/topics/manifest/activity-element#lmode may also be relevant here.

**Comment by mvvvv** [comment]
I'm experiencing this issue with OpenXR on the Meta Quest. Your thread explains well why Meta discourages exit by the application and asks to systematically use the "Meta" button which pauses and allows Resume or Quit.
My problem is that not all other platforms have this button.

**Comment by paulzhang5511** [comment]
#https://github.com/rust-windowing/winit/issues/3948

**Comment by SolidTux** [comment]
I'm also running into this issue when trying to get deep links to work. There you are basically guaranteed to run into this problem.

**Comment by Rustacian** [comment]
It would be great if this could be solved for Windows apps as well.
Should I create a separate issue for it?

**Comment by kchibisov** [comment]
If you need to re-run `loop`, use API that does so, like `run_on_demand`.

**Comment by Johannes0021** [comment]
I ran into a similar problem where the Android **MainEvent::Destroy** event didn't properly exit the
event loop, preventing **android_main()** from returning.

With a workaround I managed to make **android_main()** return, but if the app wasn't swiped away
from recent apps, it leads to the issue described in this thread.

What worked for me was calling **System.exit(0)** via JNI to fully terminate the process.
Not sure if that's the recommended approach, but it solved the problem for me.

Related issue: https://github.com/rust-windowing/winit/issues/4303

```rust
#[no_mangle]
fn android_main(android_app: AndroidApp) {
    // Do winit stuff...

    let vm = unsafe { jni::JavaVM::from_raw(android_app.vm_as_ptr() as *mut jni::sys::JavaVM) }
        .expect("Failed to create JavaVM from raw pointer!");

    let mut env = vm.attach_current_thread()
        .expect("Failed to attach current thread to JavaVM!");

    let system_class = env.find_class("java/lang/System")
        .expect("Failed to find java/lang/System class!");

    env.call_static_method(system_class, "exit", "(I)V", &[jni::objects::JValue::Int(0)])
        .expect("Failed to call System.exit(0)!");
}
```

**Comment by MarijnS95** [comment]
Not returning from  `android_main()` causes `android-activity` to not be able to `finish()` the `Activity` (in Native and mostly Java land):

https://github.com/rust-mobile/android-activity/blob/ac2e17e977a61f7753099b80500ded4c84061bae/android-activity/src/native_activity/glue.rs#L880-L899

**Comment by Johannes0021** [comment]
Ah, I see now.

**Comment by rib** [comment]
Skimming this issue - it's possible that there's more than one issue being discussed.

Since the original example is based on Winit 0.29 and based on the `run()` API that used to call `std::process::exit()` on Android, that may explain some of the issues hit, due to bypassing a clean Android Activity lifecycle.

Winit 0.29 used to assume that the running event loop effectively had ownership over the full process which was not a fair assumption on Android.

That particular issue should be addressed in winit 0.30.x, since the Even Loop 3.0 changes landed.

With Winit 0.30 on Android it should work to run and exit one event loop and then run and exit a second event loop.

Ref: https://github.com/rust-windowing/winit/issues/2709 and  https://github.com/rust-windowing/winit/pull/2767

There are still other design limitations with android-activity (stemming from NativeActivity and GameActivity) that mean it's not currently possible to run multiple Activities in a single Rust process (which the discussion above touches on too) but it would be good to differentiate that from any older lifecycle issues caused by `std::process::exit()` in Winit 0.29.

Is there maybe some regression in 0.30 that's re-introduced the possibility of `std::process::exit()` being called by Winit?

**Comment by MarijnS95** [comment]
In addition, I may remember this incorrectly but it seemed that in some scenarios (calling `exit()` with a nonzero code, and/or never handling certain events) leads to Android repeatedly restarting the application with the pending `Intent`, which subsequently crashes andsoforth.

FWIW it's perfectly possible to support/run multiple activities with at least `NativeActivity`, this has never been a limitation of Android but implementation deficits in the 3 (yes, *three*!) Android Activity abstraction crates that `winit` already went through. I've been working on refactoring `android-activity` to support that, is anyone still interested in seeing this soon-ish?

Does the `winit-android` backend split finally allow us to release breaking changes to this backend without affecting the rest of `winit`?

**Comment by rib** [comment]
> FWIW it's perfectly possible to support/run multiple activities with at least NativeActivity, this has never been a limitation of Android 

It's been a while since I looked at it but my vague recollection was that in the end NativeActivity's design (not Android's) and also GameActivity both make assumptions that there is only one Activity that they are managing.

So even though Android supports multiple Activities, and you can combine one NativeActivity with multiple non-native Activities, I was fairly sure we came to the conclusion before that we would need to create our own RustActivity if we wanted to support multiple "native" Activities.

> but implementation deficits in the 3 (yes, three!) Android Activity abstraction crates that winit already went through.

This feels a little disingenuous. When android-activity was first implemented there was a notable effort to ensure the implementation avoided any static state so that there wouldn't be a fundamental blocker to allowing multiple native activities if the underlying NativeActivity class allows it.

At the end of the day though android-activity was intentionally implemented to match the design of Google's android_native_app_glue.c (it was originally based on their C code) which is the official counterpart to NativeActivity.

The overall design of NativeActivity certainly has its limitations but afik the Rust implementation should support the same things as the C implementation.

As far as I know, if there's any practical way to instantiate multiple NativeActivities that I'd guess that's also technically somehow possible with android-activity, but considering how they are designed to hand out surfaces for rendering fullscreen apps I'm not sure how practical it would be.

Maybe one thing that comes to mind is that the ndk-context crate's design is based around sharing global static state, including a reference to the Activity. android-activity fills in the state for ndk-context but in itself android-activity doesn't depend on ndk-context and isn't oppinionated about whether applications or other crates depend on ndk-context (and they could instead get the state from the AndroidApp I guess - or rely on the fact that android-activity attaches the thread to the JVM before calling android_main).

> I've been working on refactoring android-activity to support that, is anyone still interested in seeing this soon-ish?

The vague pipe dream for some time has been to implement a RustActivity that would help us address limitations in NativeActivity / GameActivity - including hopefully making it practical to support multiple activities.

A couple of times I've briefly poked at this but never implemented anything interesting enough to share.

Yeah, I'd be interested to see what you have implemented.

I'm surprised we don't have an android-activity issue open about supporting multiple activities - I guess over the years the discussion has been scattered all over the place and I can't recall where we last discussed it in more detail.


Thinking out aloud - in Winit it could make sense to keep open the option of having two Android backends:
- One for "standalone" native Rust applications based on an Activity like NativeActivity or GameActivity that is geared towards applications with a dedicated native Rust thread that talks to the main Java thread via IPC and are probably geared towards rendering fullscreen.
- .. and a second for applications whose Rust code is loaded as a library via some Activity without spawning a native thread and are driven by events off the main Java thread without having a main loop in Rust (similar to Winit's web backend)

These are two broadly different architectures that are both valid for different use cases and I could see value in being able to use Winit in both cases with different trade offs.

I'm not sure off the top of my head how practical it would be to support both models in the android-activity crate directly.

... anyway, sorry for going off topic.
